### PR TITLE
[mlir][vector] Add distribution pattern for vector.create_mask

### PR DESCRIPTION
**Pull request by qedawkins**
This is the last step needed for basic support for distributing masked vector code. The lane id gets delinearized based on the distributed mask shape and then compared against the original mask sizes to compute the bounds for the distributed mask. Note that the distribution of masks is implicit on the shape specified by the warp op. As a result, it is the responsibility of the consumer of the mask to ensure the distributed mask will match its own distribution semantics.

### CODE REVIEW

File: mlir/lib/Dialect/Vector/Transforms/VectorDistribute.cpp [hunk 1]
```diff
@@ -1047,6 +1047,82 @@ struct WarpOpShapeCast : public OpRewritePattern<WarpExecuteOnLane0Op> {
...
+    assert(!delinearizedIds.empty());
+
+    AffineExpr s0, s1;
+    bindSymbols(rewriter.getContext(), s0, s1);
+    SmallVector<Value> newOperands;
+    for (int i = 0, e = distShape.size(); i < e; ++i) {
+      // Get `mask_size[i] - lane_id[i] * (seq_sizes[i]/dist_sizes[i])` to find
+      // the distance from the largest mask index owned by this lane to the
+      // original mask size. vector.create_mask implicitly clamps mask sizes to
+      // the range [0, mask_vector_size[i]].
```
Comment (qedawkins) [inline mlir/lib/Dialect/Vector/Transforms/VectorDistribute.cpp:1112]: I reread the semantics for create_mask, and values are implicitly clamped to a valid range, in contrast to constant_mask which requires the constant values to be within [0, mask_size], as well as zero sizes must propagate through the whole mask. This lets me drop the explicit clamping I was doing before.

File: mlir/lib/Dialect/Vector/Transforms/VectorDistribute.cpp [hunk 2]
```diff
@@ -1071,6 +1071,82 @@ struct WarpOpShapeCast : public OpRewritePattern<WarpExecuteOnLane0Op> {
...
+          return warpOp.isDefinedOutsideOfRegion(value);
+        }))
+      return failure();
+
+    Location loc = mask.getLoc();
+    unsigned operandIndex = yieldOperand->getOperandNumber();
+
+    auto distType = cast<VectorType>(warpOp.getResult(operandIndex).getType());
+    VectorType seqType = mask.getVectorType();
+    auto seqShape = seqType.getShape();
```
Comment (antiagainst) [inline mlir/lib/Dialect/Vector/Transforms/VectorDistribute.cpp:1115]: Nit: Prefer to spell out the type if it's not directly visible from the statement.

File: mlir/lib/Dialect/Vector/Transforms/VectorDistribute.cpp [hunk 3]
```diff
@@ -1071,6 +1071,82 @@ struct WarpOpShapeCast : public OpRewritePattern<WarpExecuteOnLane0Op> {
...
+                           delinearizedIds))
+      return rewriter.notifyMatchFailure(
+          mask, "cannot delinearize lane ID for distribution");
+    assert(!delinearizedIds.empty());
+
+    AffineExpr s0, s1;
+    bindSymbols(rewriter.getContext(), s0, s1);
+    SmallVector<Value> newOperands;
+    for (int i = 0, e = distShape.size(); i < e; ++i) {
+      // Get `mask_size[i] - lane_id[i] * (seq_sizes[i]/dist_sizes[i])` to find
```
Comment (antiagainst) [inline mlir/lib/Dialect/Vector/Transforms/VectorDistribute.cpp:1133]: The comment seems wrong to me. This should be `mask_dim_range_upper_limit[i] - lane[i] * distributed_shape[i]`?

File: mlir/lib/Dialect/Vector/Transforms/VectorDistribute.cpp [hunk 4]
```diff
@@ -1071,6 +1071,82 @@ struct WarpOpShapeCast : public OpRewritePattern<WarpExecuteOnLane0Op> {
...
+      // the distance from the largest mask index owned by this lane to the
+      // original mask size. vector.create_mask implicitly clamps mask sizes to
+      // the range [0, mask_vector_size[i]].
```
Comment (antiagainst) [inline mlir/lib/Dialect/Vector/Transforms/VectorDistribute.cpp:1136]: The range is `[..)` actually.

File: mlir/lib/Dialect/Vector/Transforms/VectorDistribute.cpp (hunk 1 above)
Comment (antiagainst) [reply mlir/lib/Dialect/Vector/Transforms/VectorDistribute.cpp:1136]: Neat!

**Review by antiagainst — APPROVED** [review]

**Review by antiagainst — APPROVED** [review]
LGTM with comments addressed.

File: mlir/lib/Dialect/Vector/Transforms/VectorDistribute.cpp (hunk 4 above)
Comment (qedawkins) [reply mlir/lib/Dialect/Vector/Transforms/VectorDistribute.cpp:1136]: I was using `mask sizes` in reference to the operands to the create_mask, not the size of the mask itself. The docs for these ops are kind of subtle about this and I don't think I'm using very formal terms here. I'll make the comment clearer.

**Pull request merged**